File: TreapTree.cpp
#include <iostream>
#include <unordered_map>
#include <list>
#include <vector>
#include <stdlib.h>
#include <time.h>


using namespace std;

class TreapTree{

	struct Node{
		Node *left;
		Node *right;
		int val;
		int count=1;
		int priority;
		
		Node(Node* l, Node *r,int v,int p): left(l),right(r),val(v),priority(p) {
		}
	};
	Node* nullNode;
	Node* root;
	
	
	void insert(int x, Node* &t){
		if(t==nullNode)	{
			t = new Node(nullNode,nullNode,x,rand());
		}else{
			if(t->val == x){
				t->count +=1;
			}else if(x<t->val){
				insert(x,t->left);
				if(t->left->priority<t->priority)
					rotateWithLeft(t);
			}else {
				insert(x,t->right);
				if(t->right->priority<t->priority)
					rotateWithRight(t);
			}
		}
	}
		
	void remove(int x,Node* &t){
		if(t!=nullNode){
			if(x<t->val){
				remove(x,t->left);
			}else if(x>t->val){
				remove(x,t->right);
			}else{
				if(t->count>1){
					t->count -= 1;
				}else{
					if(t->left->priority<t->right->priority){
						rotateWithLeft(t);
					}else{
						rotateWithRight(t);
					}
					
					if(t!=nullNode){
						remove(x,t);
					}else{
						delete t->left;
						t->left = nullNode;
					}
				}
			}
		}
	}
	
	void rotateWithLeft(Node* &t){
		Node *l = t->left;
		t->left = l->right;
		l->right = t;
		t = l;
	}
	
	void rotateWithRight(Node* &t){
		Node *r = t->right;
		t->right = r->left;
		r->left = t;
		t = r;
	}
	
	void print(Node* t,string s){
		if(t==nullNode)
			return ;
		print(t->left,s+"\t");
		cout<<s<<t->val<<","<<t->count<<endl;
		print(t->right,s+"\t");
	}
	
public:	
	TreapTree(){
		nullNode = new Node(NULL,NULL,0,INT_MAX);
		nullNode->left = nullNode->right = nullNode;
		root = nullNode;
	}
	void insert(int x){
		insert(x,root);
	}
	
	void remove(int x){
		remove(x,root);
	}
	
	void print(){
		print(root,"");
	}
	
	
};

/*
class Solution {
public:
    vector<int> countSmaller(vector<int>& nums) {
        
    }
};*/

int main(){
	srand(time(NULL));
	
	int arr[] = {5,7,2,6,7,9,10,1,5,6,7,3,9};
	int n = sizeof(arr)/sizeof(arr[0]);
	TreapTree tree;
	for(int i=0;i<n;++i){
		tree.insert(arr[i]);
	}
	tree.print();
	
	tree.remove(5);
	tree.print();
	tree.remove(5);
	tree.print();
	tree.remove(3);
	tree.print();
	return 0;
}

File: TreapTree.py
import random
from sys import maxint

class Node:
  def __init__(self,left,right,val,prio):
    self.left = left
    self.right = right
    self.val = val
    self.count = 1
    self.prio = prio

class TreapTree:
  '''
  ### Implementation of Treap Tree
  '''
  def rand(self):
  	return random.randint(1,maxint-1)

  def __init__(self):
    random.seed()   
    self.nil = Node(None,None,0,maxint)
    self.nil.left = self.nil.right = self.nil
    self.root = self.nil

  def insert(self,x):
    self.root = self._insert(x,self.root)

  def remove(self,x):
    self.root = self._remove(x,self.root)

  def display(self):
    self._display(self.root,'')

  def _insert(self,x,node):
    if node is self.nil:
      return Node(self.nil,self.nil,x,self.rand())
    else:
      if x == node.val:
      	node.count += 1
      elif x < node.val:
      	node.left = self._insert(x,node.left)
      	if node.left.prio < node.prio:
      	  node = self.rotateWithLeft(node)
      else:
      	node.right = self._insert(x,node.right)
      	if node.right.prio < node.prio:
          node = self.rotateWithRight(node)
      return node   
		
  def _remove(self,x,node):
    if node is self.nil:
      return node
    else:
      if x < node.val:
        node.left = self._remove(x,node.left)
      elif x > node.val:
        node.right = self._remove(x,node.right)
      else:
        if node.count > 1:
          node.count -= 1
        else:
          if node.left.prio < node.right.prio:
            node = self.rotateWithLeft(node)
          else:
            node = self.rotateWithRight(node)

          if not (node is self.nil):
            node = self._remove(x,node)
          else:
            node.left = self.nil
      return node
 
  def rotateWithLeft(self,node):
  	tmp = node.left
  	node.left = tmp.right
  	tmp.right = node
  	return tmp

  def rotateWithRight(self,node):
  	tmp = node.right
  	node.right = tmp.left
  	tmp.left = node
  	return tmp

  def _display(self,node,s):
    if node is self.nil:
      return 
    self._display(node.left,s+'\t')
    print s,node.val,node.count
    self._display(node.right,s+'\t')


if __name__ == "__main__":
  tree = TreapTree()
  data = [5,7,2,6,7,9,10,1,5,6,7,3,9]

  for x in data:
    tree.insert(x)

  tree.display()
  print '=' * 10
  tree.remove(5)
  tree.display()
  print '=' * 10
  tree.remove(5)
  tree.display()
  print '=' * 10
  tree.remove(3)
  tree.display()
